### DISCUSSION TITLE
onValue callback for props

### DISCUSSION THREAD

**Discussion by davidhernandeze** [vuejs/rfcs#681] [General]
Add the ability to trigger an action when a prop value is first set. I found this is a very common scenario when you want a component to render a loading state while waiting for the parent to pass the first value of a prop. The child component receives a null value at first, and waits for the prop to change to render its loaded state. If the component receives a not null value from the beggining, the callback is called after onMounted hook.

```js
const props = defineProps({
  products: {
    required: true,
    onValue: onProductsLoad
  }
})

const isLoading = ref(true)

function onProductsLoad() { 
  isLoading.value = false 
}
```

**Comment by nickmessing** [comment]
You can have a tiny utility function for that:

```js
import { watch, toRef } from "vue";

export function onFirstValue(something, fn) {
  const somethingRef = toRef(something)

  if (somethingRef.value != null) {
    fn(somethingRef.value)
    return
  }

  const stopWatching = watch(somethingRef, newValue => {
    if (newValue != null) {
      stopWatching()
      fn(newValue)
    }
  })
}
```

And use it like:
```js
const props = defineProps(['myProp'])

onFirstValue(
  () => props.myProp,
  firstValue => {
    console.log(firstValue)
  }
)
```

[Example in action](https://play.vuejs.org/#eNqFVE2P0zAQ/SuDL0mlKjmUU0krPrRIcIDVLoID5lDSSZtdx7Zspy2q8t8Z222SLmX3EsUzb57fvJnkyN5pne1aZHNW2NLU2oFF1+oll3WjlXFwBIMVdFAZ1UBC0ITLPvlBNfqUyXJ/8FwBUCppHWij9PeVaBEWniadcFnk8R66gQ4OGy1WDukEUPxunVMS3paiLh8XnI3LE4fWJZwt7zHywi5knALOfI6zIo8EL5PNrvLMLglCc/Pmzy2BxvWcQU6IIh+JZ1PmLPVc1ZvswSpJfh49CWclsdQCzVftavKEszmEjM+thFD7zyHmTIvTc7zcYvl4Jf5gDz7G2a1Bi2ZHUvqcW5kNkgk+fXP/BQ/03icbtW4FoZ9J3qFVovUaI+x9K9cke4QLaj+Fwddy883eHBxKe27KC/XILuA5o0XwBv6v9UHuLHsd6rjsyMXzEj23kEp+rI11cZjnzczycTh5c7mEloa+xqqW6Kdp059JnGvyi1aSy3Fp6tWkE1gsY2UWkaGtariX0qdm/C1KYCbUJh0AxBuaemHltcHl8XjS2HVF7gP/LtdYIC0YudO7sV+5cjul/b0bvtNgP2fBBDwEYNXK0o/hwr3UqgbdlsY5hUpOYkfRtT7jaReRfoAH1wDqCoYYAbL4Lb1agGyFOPGRbfIKKhgE9FtwrQlkZNboeqf0D98ZldD1ockLkilI3D8dhdfTh5+qgAtS/y+KUVJ3rjnFSAk96NCx7i9iA8Ao)
